pkg/clusteragent/autoscaling/controller.go

```diff
@@ -0,0 +1,152 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+package autoscaling
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// Controller is generic controller that can be used to implement a controller for a specific resource
+type Controller struct {
+	processor Processor
+	synced    cache.InformerSynced
+	context   context.Context
+
+	// Fields available to child controllers
+	ID        string
+	Client    dynamic.Interface
+	Lister    cache.GenericLister
+	Workqueue workqueue.RateLimitingInterface
+	IsLeader  func() bool
+}
+
+// NewController returns a new workload autoscaling controller
+func NewController(
+	controllerID string,
+	processor Processor,
+	client dynamic.Interface,
+	informer dynamicinformer.DynamicSharedInformerFactory,
+	gvr schema.GroupVersionResource,
+	isLeader func() bool,
+	observable Observable,
+) (*Controller, error) {
+	mainInformer := informer.ForResource(gvr)
+	c := &Controller{
+		processor: processor,
+		ID:        controllerID,
+		Client:    client,
+		Lister:    mainInformer.Lister(),
+		synced:    mainInformer.Informer().HasSynced,
+		Workqueue: workqueue.NewRateLimitingQueue(workqueue.DefaultItemBasedRateLimiter()),
+		IsLeader:  isLeader,
+	}
+
+	if _, err := mainInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.enqueue,
+		DeleteFunc: c.enqueue,
+		UpdateFunc: func(obj, new interface{}) {
+			c.enqueue(new)
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("cannot add event handler to informer: %v", err)
+	}
+
+	// We use an observer on the store to propagate events as soon as possible
+	observable.RegisterObserver(Observer{
+		SetFunc: c.enqueueID,
+	})
+
+	return c, nil
+}
+
+// Run starts the controller to handle objects
+func (c *Controller) Run(ctx context.Context) {
+	if ctx == nil {
+		log.Errorf("Cannot run with a nil context")
+		return
+	}
+	c.context = ctx
+
+	defer c.Workqueue.ShutDown()
+
+	log.Infof("Starting controller id: %s (waiting for cache sync)", c.ID)
+	if !cache.WaitForCacheSync(ctx.Done(), c.synced) {
+		log.Errorf("Failed to wait for caches to sync for controller id: %s", c.ID)
+		return
+	}
+
+	go wait.Until(c.worker, time.Second, ctx.Done())
```

Comment:
Using wait.Until() here will execute `c.worker` every second. Is there a reason why we need this extra second ?
Would it be possible to do something similar to this ?
```
func (c* Controller) worker() {
  go func() {
    <- ctx.Done()
    c.Workqueue.Shutdown()
  }
}
for {
  item, shutdown := c.Workqueue.Get()
  if shutdown {
    // shutdown logic
  }
  // processing
  // ...
}
```